templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml

```
@@ -20,89 +20,140 @@ contents:
+                pull_baremetal_runtime_cfg_image &
+            fi
+
+            # This function is subject to a timeout. The timeout may interrupt podman run's image pull if the image is
+            # not available and cause image corruption [1]. Image should be available or being pulled before we reach
+            # this point. Wait until its available.
+            # [1] https://github.com/containers/podman/issues/14003
+            until /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; do
+                >&2 echo "NM resolv-prepender: Waiting for baremetal runtime cfg image to be available"
+                sleep 1
+            done
+
+            # Image must exist before reaching this point otherwise timeout may cause image corruption.
+            NAMESERVER_IP=$(/usr/bin/podman run --rm \
+                --authfile /var/lib/kubelet/config.json \
+                --net=host \
+                {{ .Images.baremetalRuntimeCfgImage }} \
+                node-ip \
+                show \
+                --retry-on-failure \
+                "{{ onPremPlatformAPIServerInternalIP . }}" \
+                "{{ onPremPlatformIngressIP . }}")
+            DOMAINS="${IP4_DOMAINS} ${IP6_DOMAINS} {{.DNS.Spec.BaseDomain}}"
+            if [[ -n "$NAMESERVER_IP" ]]; then
+                if systemctl -q is-enabled systemd-resolved; then
+                    >&2 echo "NM resolv-prepender: Configure for OKD domain and local IP"
                     mkdir -p /etc/systemd/resolved.conf.d
-                    echo "[Resolve]" > /etc/systemd/resolved.conf.d/60-kni.conf
-                    echo "DNS=$NAMESERVER_IP" >> /etc/systemd/resolved.conf.d/60-kni.conf
-                    echo "Domains=$DOMAINS" >> /etc/systemd/resolved.conf.d/60-kni.conf
-                    if systemctl -q is-active systemd-resolved; then
-                        >&2 echo "NM resolv-prepender: restarting systemd-resolved"
-                        systemctl restart systemd-resolved
+                    KNICONFTMPPATH="$(mktemp)"
+                    KNICONFPATH="/etc/systemd/resolved.conf.d/60-kni.conf"
+                    KNICONFDONEPATH="/run/resolv-prepender-kni-conf-done"
+                    echo "[Resolve]" > "${KNICONFTMPPATH}"
+                    echo "DNS=$NAMESERVER_IP" >> "${KNICONFTMPPATH}"
+                    echo "Domains=$DOMAINS" >> "${KNICONFTMPPATH}"
+
+                    # If KNI conf is not created or doesn't match what is generated or
+                    # if we haven't completed a full update - create or update it
+                    if [[ ! -f "${"KNICONFPATH"} ]] || [[ ! -f "${"KNICONFDONEPATH"} ]] || ! cmp --silent "${KNICONFPATH}" "${KNICONFTMPPATH}"; then
+                        >&2 echo "NM resolv-prepender: Creating/updating /etc/systemd/resolved.conf.d/60-kni.conf"
+                        rm -f "${KNICONFDONEPATH}"
+                        mv -f "${KNICONFTMPPATH}" "${KNICONFPATH}"
```

```suggestion
                        mv -f "${KNICONFTMPPATH}" "${KNICONFPATH}"
                        chown systemd-resolve:systemd-resolve "${KNICONFPATH}"
                        chcon system_u:object_r:etc_t:s0 "${KNICONFPATH}"
```
tmp file has invalid permissions and SELinux label

---

Or simply `cat "${KNICONFTMPPATH}" > "${KNICONFPATH}"`